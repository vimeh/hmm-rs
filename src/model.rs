use indextree::NodeId as TreeNodeId;

pub type NodeId = TreeNodeId;

#[derive(Debug, Clone)]
pub struct Node {
    pub title: String,
    pub is_collapsed: bool,
    pub is_hidden: bool,
}

impl Node {
    pub fn new(title: String) -> Self {
        Self {
            title,
            is_collapsed: false,
            is_hidden: false,
        }
    }

    pub fn is_hidden(&self) -> bool {
        self.is_hidden || self.title.starts_with("[HIDDEN] ")
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_node_creation() {
        let node = Node::new("Test Node".to_string());
        assert_eq!(node.title, "Test Node");
        assert!(!node.is_collapsed);
        assert!(!node.is_hidden);
    }

    #[test]
    fn test_is_hidden_with_flag() {
        let mut node = Node::new("Normal Node".to_string());
        assert!(!node.is_hidden());

        node.is_hidden = true;
        assert!(node.is_hidden());
    }

    #[test]
    fn test_is_hidden_with_title_prefix() {
        let node = Node::new("[HIDDEN] Secret Node".to_string());
        assert!(node.is_hidden());
    }

    #[test]
    fn test_is_hidden_with_both() {
        let mut node = Node::new("[HIDDEN] Secret".to_string());
        node.is_hidden = true;
        assert!(node.is_hidden());
    }

    #[test]
    fn test_node_with_unicode_title() {
        let node = Node::new("âœ“ Task Complete ðŸŽ¯".to_string());
        assert_eq!(node.title, "âœ“ Task Complete ðŸŽ¯");
    }
}
