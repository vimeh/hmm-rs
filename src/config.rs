use anyhow::Result;
use clap::Parser;
use directories::ProjectDirs;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Parser, Debug)]
#[command(name = "h-m-m")]
#[command(version, about = "A simple, fast, keyboard-centric terminal-based tool for working with mind maps", long_about = None)]
pub struct CliArgs {
    /// The mind map file to open
    pub filename: Option<PathBuf>,

    /// Custom configuration file path
    #[arg(long)]
    pub config: Option<PathBuf>,

    /// Debug configuration
    #[arg(long)]
    pub debug_config: bool,

    /// Initial depth to expand nodes
    #[arg(long)]
    pub initial_depth: Option<usize>,

    /// Show hidden nodes
    #[arg(long)]
    pub show_hidden: Option<bool>,

    /// Auto-save mode
    #[arg(long)]
    pub auto_save: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppConfig {
    #[serde(default = "default_max_parent_node_width")]
    pub max_parent_node_width: usize,

    #[serde(default = "default_max_leaf_node_width")]
    pub max_leaf_node_width: usize,

    #[serde(default = "default_line_spacing")]
    pub line_spacing: usize,

    #[serde(default = "default_symbol1")]
    pub symbol1: String,

    #[serde(default = "default_symbol2")]
    pub symbol2: String,

    #[serde(default = "default_show_hidden")]
    pub show_hidden: bool,

    #[serde(default = "default_initial_depth")]
    pub initial_depth: usize,

    #[serde(default = "default_center_lock")]
    pub center_lock: bool,

    #[serde(default = "default_focus_lock")]
    pub focus_lock: bool,

    #[serde(default = "default_max_undo_steps")]
    pub max_undo_steps: usize,

    #[serde(default = "default_auto_save")]
    pub auto_save: bool,

    #[serde(default = "default_auto_save_interval")]
    pub auto_save_interval: usize,

    #[serde(default = "default_echo_keys")]
    pub echo_keys: bool,

    #[serde(default = "default_post_export_command")]
    pub post_export_command: String,

    #[serde(default = "default_clipboard")]
    pub clipboard: ClipboardType,

    #[serde(default = "default_clipboard_file")]
    pub clipboard_file: PathBuf,

    #[serde(default)]
    pub clipboard_in_command: String,

    #[serde(default)]
    pub clipboard_out_command: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ClipboardType {
    Os,
    File,
    Command,
}

impl Default for AppConfig {
    fn default() -> Self {
        Self {
            max_parent_node_width: default_max_parent_node_width(),
            max_leaf_node_width: default_max_leaf_node_width(),
            line_spacing: default_line_spacing(),
            symbol1: default_symbol1(),
            symbol2: default_symbol2(),
            show_hidden: default_show_hidden(),
            initial_depth: default_initial_depth(),
            center_lock: default_center_lock(),
            focus_lock: default_focus_lock(),
            max_undo_steps: default_max_undo_steps(),
            auto_save: default_auto_save(),
            auto_save_interval: default_auto_save_interval(),
            echo_keys: default_echo_keys(),
            post_export_command: default_post_export_command(),
            clipboard: default_clipboard(),
            clipboard_file: default_clipboard_file(),
            clipboard_in_command: String::new(),
            clipboard_out_command: String::new(),
        }
    }
}

fn default_max_parent_node_width() -> usize {
    25
}
fn default_max_leaf_node_width() -> usize {
    55
}
fn default_line_spacing() -> usize {
    1
}
fn default_symbol1() -> String {
    "✓".to_string()
}
fn default_symbol2() -> String {
    "✗".to_string()
}
fn default_show_hidden() -> bool {
    false
}
fn default_initial_depth() -> usize {
    1
}
fn default_center_lock() -> bool {
    false
}
fn default_focus_lock() -> bool {
    false
}
fn default_max_undo_steps() -> usize {
    24
}
fn default_auto_save() -> bool {
    false
}

fn default_auto_save_interval() -> usize {
    30 // 30 seconds default
}
fn default_echo_keys() -> bool {
    false
}
fn default_post_export_command() -> String {
    String::new()
}
fn default_clipboard() -> ClipboardType {
    ClipboardType::Os
}
fn default_clipboard_file() -> PathBuf {
    PathBuf::from("/tmp/h-m-m")
}

pub fn load_config(args: &CliArgs) -> Result<AppConfig> {
    let mut config = config::Config::builder();

    // Start with default values
    config = config.add_source(config::Config::try_from(&AppConfig::default())?);

    // Try to load from config file
    let config_path = if let Some(ref path) = args.config {
        path.clone()
    } else {
        get_default_config_path()
    };

    if config_path.exists() {
        config = config.add_source(config::File::from(config_path));
    }

    // Apply environment variables prefixed with HMM_
    config = config.add_source(config::Environment::with_prefix("HMM").separator("_"));

    // Apply command line arguments
    if let Some(depth) = args.initial_depth {
        config = config.set_override("initial_depth", depth as i64)?;
    }
    if let Some(show) = args.show_hidden {
        config = config.set_override("show_hidden", show)?;
    }
    if let Some(auto) = args.auto_save {
        config = config.set_override("auto_save", auto)?;
    }

    let config = config.build()?;
    Ok(config.try_deserialize()?)
}

fn get_default_config_path() -> PathBuf {
    if let Some(proj_dirs) = ProjectDirs::from("", "", "h-m-m") {
        let config_dir = proj_dirs.config_dir();
        config_dir.join("h-m-m.conf")
    } else {
        // Fallback to home directory
        let home = std::env::var("HOME").unwrap_or_else(|_| ".".to_string());
        PathBuf::from(home)
            .join(".config")
            .join("h-m-m")
            .join("h-m-m.conf")
    }
}
