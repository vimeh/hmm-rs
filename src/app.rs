use crate::config::AppConfig;
use crate::model::{Node, NodeId};
use indextree::Arena;
use std::path::PathBuf;

#[derive(Debug, Clone, PartialEq)]
pub enum AppMode {
    Normal,
    Editing { buffer: String, cursor_pos: usize },
    Search { query: String },
    Help,
}

pub struct AppState {
    pub running: bool,
    pub mode: AppMode,
    pub tree: Arena<Node>,
    pub root_id: Option<NodeId>,
    pub active_node_id: Option<NodeId>,
    pub config: AppConfig,
    pub filename: Option<PathBuf>,

    // Viewport state
    pub viewport_top: f64,
    pub viewport_left: f64,
    pub terminal_width: u16,
    pub terminal_height: u16,

    // Undo/Redo history
    pub history: Vec<Arena<Node>>,
    pub history_index: usize,

    // Message for status line
    pub message: Option<String>,

    // Search state
    pub search_results: Vec<NodeId>,
    pub search_index: usize,

    // Clipboard
    pub clipboard: Option<String>,
}

impl AppState {
    pub fn new(config: AppConfig) -> Self {
        let tree = Arena::new();

        Self {
            running: true,
            mode: AppMode::Normal,
            tree,
            root_id: None,
            active_node_id: None,
            config,
            filename: None,
            viewport_top: 0.0,
            viewport_left: 0.0,
            terminal_width: 80,
            terminal_height: 24,
            history: Vec::new(),
            history_index: 0,
            message: None,
            search_results: Vec::new(),
            search_index: 0,
            clipboard: None,
        }
    }

    pub fn push_history(&mut self) {
        // Remove any redo history
        self.history.truncate(self.history_index);

        // Add current state to history
        self.history.push(self.tree.clone());
        self.history_index += 1;

        // Limit history size
        if self.history.len() > self.config.max_undo_steps {
            self.history.remove(0);
            self.history_index -= 1;
        }
    }

    pub fn undo(&mut self) -> bool {
        if self.history_index > 0 {
            self.history_index -= 1;
            self.tree = self.history[self.history_index].clone();
            true
        } else {
            false
        }
    }

    pub fn redo(&mut self) -> bool {
        if self.history_index < self.history.len() - 1 {
            self.history_index += 1;
            self.tree = self.history[self.history_index].clone();
            true
        } else {
            false
        }
    }

    pub fn set_message(&mut self, msg: impl Into<String>) {
        self.message = Some(msg.into());
    }

    pub fn clear_message(&mut self) {
        self.message = None;
    }
}
