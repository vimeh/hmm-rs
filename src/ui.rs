use crate::app::{AppMode, AppState};
use crate::layout::LayoutEngine;
use crate::model::NodeId;
use ratatui::{
    Frame,
    layout::{Constraint, Direction, Layout, Rect},
    style::{Color, Modifier, Style},
    text::{Line, Span},
    widgets::{Block, Borders, Paragraph, Wrap},
};

pub fn render(frame: &mut Frame, app: &mut AppState) {
    // Update terminal size
    let size = frame.area();
    app.terminal_width = size.width;
    app.terminal_height = size.height;

    // Calculate layout
    let layout = LayoutEngine::calculate_layout(app);

    // Create main layout chunks
    let chunks = Layout::default()
        .direction(Direction::Vertical)
        .constraints([Constraint::Min(1), Constraint::Length(1)])
        .split(size);

    // Render based on mode
    match &app.mode {
        AppMode::Help => render_help(frame, chunks[0]),
        _ => render_mind_map(frame, app, &layout, chunks[0]),
    }

    // Render status/input line
    render_status_line(frame, app, chunks[1]);
}

fn render_mind_map(frame: &mut Frame, app: &AppState, layout: &LayoutEngine, area: Rect) {
    // Create a buffer for the mind map
    let mut map_buffer: Vec<Vec<char>> = vec![vec![' '; area.width as usize]; area.height as usize];
    let mut style_buffer: Vec<Vec<Style>> =
        vec![vec![Style::default(); area.width as usize]; area.height as usize];

    // Draw connections
    draw_connections(&mut map_buffer, app, layout, area);

    // Draw nodes
    draw_nodes(&mut map_buffer, &mut style_buffer, app, layout, area);

    // Convert buffer to text
    let mut lines = Vec::new();
    for (y, row) in map_buffer.iter().enumerate() {
        let mut spans = Vec::new();
        let mut current_style = Style::default();
        let mut current_text = String::new();

        for (x, &ch) in row.iter().enumerate() {
            let style = style_buffer[y][x];
            if style != current_style {
                if !current_text.is_empty() {
                    spans.push(Span::styled(current_text.clone(), current_style));
                    current_text.clear();
                }
                current_style = style;
            }
            current_text.push(ch);
        }

        if !current_text.is_empty() {
            spans.push(Span::styled(current_text, current_style));
        }

        lines.push(Line::from(spans));
    }

    let paragraph = Paragraph::new(lines);
    frame.render_widget(paragraph, area);
}

fn draw_connections(
    buffer: &mut Vec<Vec<char>>,
    app: &AppState,
    layout: &LayoutEngine,
    area: Rect,
) {
    if let Some(root_id) = app.root_id {
        draw_node_connections(buffer, app, layout, root_id, area);
    }
}

fn draw_node_connections(
    buffer: &mut Vec<Vec<char>>,
    app: &AppState,
    layout: &LayoutEngine,
    node_id: NodeId,
    area: Rect,
) {
    let node = app.tree.get(node_id).unwrap().get();

    if let Some(node_layout) = layout.nodes.get(&node_id) {
        let children: Vec<NodeId> = if !app.config.show_hidden {
            node_id
                .children(&app.tree)
                .filter(|cid| {
                    let child = app.tree.get(*cid).unwrap().get();
                    !child.is_hidden()
                })
                .collect()
        } else {
            node_id.children(&app.tree).collect()
        };

        if node.is_collapsed && !children.is_empty() {
            // Draw collapsed indicator
            let x = (node_layout.x + node_layout.w + 1.0 - app.viewport_left) as usize;
            let y = (node_layout.y + node_layout.yo - app.viewport_top) as usize;

            if x < area.width as usize && y < area.height as usize {
                draw_text(buffer, x, y, " [+]");
            }
        } else if !children.is_empty() {
            // Draw connections to children
            let parent_x = (node_layout.x + node_layout.w - app.viewport_left) as usize;
            let parent_y = (node_layout.y + node_layout.yo + node_layout.lh / 2.0
                - app.viewport_top) as usize;

            // Draw horizontal line from parent node text
            if parent_x < area.width as usize && parent_y < area.height as usize {
                for x in parent_x..(parent_x + 4).min(area.width as usize) {
                    set_char(buffer, x, parent_y, '─');
                }
            }

            // Calculate vertical connector position (right after parent's horizontal line)
            let conn_x = parent_x + 4;

                // Get first and last child positions
                if let (Some(first_layout), Some(last_layout)) =
                    (layout.nodes.get(&children[0]), layout.nodes.get(&children[children.len() - 1])) {

                    let first_y = (first_layout.y + first_layout.yo + first_layout.lh / 2.0 - app.viewport_top) as usize;
                    let last_y = (last_layout.y + last_layout.yo + last_layout.lh / 2.0 - app.viewport_top) as usize;

                    // Draw continuous vertical line from parent level to last child
                    if conn_x < area.width as usize {
                        // Connect from parent y to first child
                        if parent_y < first_y {
                            for y in parent_y..first_y {
                                if y < area.height as usize {
                                    set_char(buffer, conn_x, y, '│');
                                }
                            }
                        }
                        // Draw through all children
                        for y in first_y..=last_y.min(area.height as usize - 1) {
                            if y < area.height as usize {
                                set_char(buffer, conn_x, y, '│');
                            }
                        }
                    }
                }

            // Now draw horizontal connectors for each child
            for (i, child_id) in children.iter().enumerate() {
                if let Some(child_layout) = layout.nodes.get(child_id) {
                    let child_x = (child_layout.x - app.viewport_left) as usize;
                    let child_y = (child_layout.y + child_layout.yo + child_layout.lh / 2.0
                        - app.viewport_top) as usize;

                    if child_y < area.height as usize && conn_x < area.width as usize {
                        // Draw horizontal line from vertical connector to just before child text
                        // Leave 2 spaces before the child text for the "──" prefix
                        if child_x > 2 {
                            for x in conn_x..(child_x - 2).min(area.width as usize) {
                                set_char(buffer, x, child_y, '─');
                            }
                            // Add the "──" prefix right before the child text
                            if child_x >= 2 {
                                set_char(buffer, child_x - 2, child_y, '─');
                                set_char(buffer, child_x - 1, child_y, '─');
                            }
                        }

                        // Place junction character at vertical line
                        if children.len() == 1 {
                            // Single child - horizontal line continues
                            set_char(buffer, conn_x, child_y, '─');
                        } else if i == 0 {
                            // First child of multiple
                            set_char(buffer, conn_x, child_y, '├');
                        } else if i == children.len() - 1 {
                            // Last child
                            set_char(buffer, conn_x, child_y, '╰');
                        } else {
                            // Middle children
                            set_char(buffer, conn_x, child_y, '├');
                        }
                    }

                    // Recursively draw connections for children
                    if !node.is_collapsed {
                        draw_node_connections(buffer, app, layout, *child_id, area);
                    }
                }
            }
        }
    }
}

fn draw_nodes(
    buffer: &mut Vec<Vec<char>>,
    style_buffer: &mut Vec<Vec<Style>>,
    app: &AppState,
    layout: &LayoutEngine,
    area: Rect,
) {
    if let Some(root_id) = app.root_id {
        draw_node_content(buffer, style_buffer, app, layout, root_id, area);
    }
}

fn draw_node_content(
    buffer: &mut Vec<Vec<char>>,
    style_buffer: &mut Vec<Vec<Style>>,
    app: &AppState,
    layout: &LayoutEngine,
    node_id: NodeId,
    area: Rect,
) {
    let node = app.tree.get(node_id).unwrap().get();

    if let Some(node_layout) = layout.nodes.get(&node_id) {
        let x = (node_layout.x - app.viewport_left) as usize;
        let y = (node_layout.y + node_layout.yo - app.viewport_top) as usize;

        // Determine node style
        let style = if Some(node_id) == app.active_node_id {
            Style::default()
                .fg(Color::Black)
                .bg(Color::Yellow)
                .add_modifier(Modifier::BOLD)
        } else if node.title.starts_with(&app.config.symbol1) {
            Style::default().fg(Color::Green)
        } else if node.title.starts_with(&app.config.symbol2) {
            Style::default().fg(Color::Red)
        } else if node.is_hidden() {
            Style::default().fg(Color::DarkGray)
        } else {
            Style::default()
        };

        // Draw node text
        if x < area.width as usize && y < area.height as usize {
            let lines = wrap_text(&node.title, node_layout.w as usize);
            for (i, line) in lines.iter().enumerate() {
                let line_y = y + i;
                if line_y < area.height as usize {
                    draw_styled_text(buffer, style_buffer, x, line_y, line, style);
                }
            }
        }

        // Draw children if not collapsed
        if !node.is_collapsed {
            let children: Vec<NodeId> = if !app.config.show_hidden {
                node_id
                    .children(&app.tree)
                    .filter(|cid| {
                        let child = app.tree.get(*cid).unwrap().get();
                        !child.is_hidden()
                    })
                    .collect()
            } else {
                node_id.children(&app.tree).collect()
            };

            for child_id in children {
                draw_node_content(buffer, style_buffer, app, layout, child_id, area);
            }
        }
    }
}

fn render_status_line(frame: &mut Frame, app: &AppState, area: Rect) {
    let content = match &app.mode {
        AppMode::Normal => {
            if let Some(ref msg) = app.message {
                msg.clone()
            } else {
                format!("h-m-m | {} nodes", app.tree.count())
            }
        }
        AppMode::Editing {
            buffer,
            cursor_pos: _,
        } => {
            format!("Edit: {}", buffer)
        }
        AppMode::Search { query } => {
            format!("Search: {}", query)
        }
        AppMode::Help => "Press ESC or q to close help".to_string(),
    };

    let style = match &app.mode {
        AppMode::Editing { .. } | AppMode::Search { .. } => Style::default()
            .fg(Color::Black)
            .bg(Color::Cyan)
            .add_modifier(Modifier::BOLD),
        _ => {
            if app.message.is_some() {
                Style::default()
                    .fg(Color::Black)
                    .bg(Color::Magenta)
                    .add_modifier(Modifier::BOLD)
            } else {
                Style::default().fg(Color::Gray).bg(Color::Black)
            }
        }
    };

    let paragraph = Paragraph::new(content)
        .style(style)
        .wrap(Wrap { trim: false });

    frame.render_widget(paragraph, area);
}

fn render_help(frame: &mut Frame, area: Rect) {
    let help_text = vec![
        Line::from(vec![Span::styled(
            "h-m-m Help",
            Style::default().add_modifier(Modifier::BOLD),
        )]),
        Line::from(""),
        Line::from(vec![Span::styled(
            "Navigation:",
            Style::default().add_modifier(Modifier::BOLD),
        )]),
        Line::from("  h/←  Move left (parent)"),
        Line::from("  j/↓  Move down"),
        Line::from("  k/↑  Move up"),
        Line::from("  l/→  Move right (child)"),
        Line::from("  g    Go to top"),
        Line::from("  G    Go to bottom"),
        Line::from("  m/~  Go to root"),
        Line::from(""),
        Line::from(vec![Span::styled(
            "Editing:",
            Style::default().add_modifier(Modifier::BOLD),
        )]),
        Line::from("  e/i  Edit node (append)"),
        Line::from("  E/I  Edit node (replace)"),
        Line::from("  o/⏎  Insert sibling"),
        Line::from("  O/⇥  Insert child"),
        Line::from("  d    Delete node"),
        Line::from("  D    Delete children"),
        Line::from(""),
        Line::from(vec![Span::styled(
            "View:",
            Style::default().add_modifier(Modifier::BOLD),
        )]),
        Line::from("  ␣    Toggle collapse"),
        Line::from("  v    Collapse all"),
        Line::from("  b    Expand all"),
        Line::from("  1-5  Collapse to level"),
        Line::from(""),
        Line::from(vec![Span::styled(
            "File:",
            Style::default().add_modifier(Modifier::BOLD),
        )]),
        Line::from("  s    Save"),
        Line::from("  S    Save as"),
        Line::from("  q    Quit"),
        Line::from(""),
        Line::from("Press ESC or q to close help"),
    ];

    let block = Block::default().borders(Borders::ALL).title(" Help ");

    let paragraph = Paragraph::new(help_text)
        .block(block)
        .wrap(Wrap { trim: false });

    frame.render_widget(paragraph, area);
}

// Helper functions
fn set_char(buffer: &mut Vec<Vec<char>>, x: usize, y: usize, ch: char) {
    if y < buffer.len() && x < buffer[y].len() {
        buffer[y][x] = ch;
    }
}

fn draw_text(buffer: &mut Vec<Vec<char>>, x: usize, y: usize, text: &str) {
    for (i, ch) in text.chars().enumerate() {
        set_char(buffer, x + i, y, ch);
    }
}

fn draw_styled_text(
    buffer: &mut Vec<Vec<char>>,
    style_buffer: &mut Vec<Vec<Style>>,
    x: usize,
    y: usize,
    text: &str,
    style: Style,
) {
    for (i, ch) in text.chars().enumerate() {
        if y < buffer.len() && x + i < buffer[y].len() {
            buffer[y][x + i] = ch;
            style_buffer[y][x + i] = style;
        }
    }
}

fn wrap_text(text: &str, max_width: usize) -> Vec<String> {
    let mut lines = Vec::new();
    let mut current_line = String::new();
    let mut current_width = 0;

    for word in text.split_whitespace() {
        let word_width = unicode_width::UnicodeWidthStr::width(word);

        if current_width > 0 && current_width + 1 + word_width > max_width {
            lines.push(current_line);
            current_line = word.to_string();
            current_width = word_width;
        } else {
            if !current_line.is_empty() {
                current_line.push(' ');
                current_width += 1;
            }
            current_line.push_str(word);
            current_width += word_width;
        }
    }

    if !current_line.is_empty() {
        lines.push(current_line);
    }

    if lines.is_empty() {
        lines.push(text.to_string());
    }

    lines
}
