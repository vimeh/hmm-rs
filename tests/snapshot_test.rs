use hmm_rs::app::{AppMode, AppState};
use hmm_rs::config::AppConfig;
use hmm_rs::model::Node;
use hmm_rs::ui;
use insta::assert_snapshot;
use ratatui::{backend::TestBackend, Terminal};

fn create_test_app_with_tree() -> AppState {
    let config = AppConfig::default();
    let mut app = AppState::new(config);

    // Create a sample tree
    let root = app.tree.new_node(Node::new("Mind Map Root".to_string()));
    let feature = app.tree.new_node(Node::new("Features".to_string()));
    let task1 = app.tree.new_node(Node::new("✓ Completed Task".to_string()));
    let task2 = app.tree.new_node(Node::new("✗ Failed Task".to_string()));
    let architecture = app.tree.new_node(Node::new("Architecture".to_string()));
    let module1 = app.tree.new_node(Node::new("model.rs".to_string()));
    let module2 = app.tree.new_node(Node::new("ui.rs".to_string()));

    root.append(feature, &mut app.tree);
    root.append(architecture, &mut app.tree);
    feature.append(task1, &mut app.tree);
    feature.append(task2, &mut app.tree);
    architecture.append(module1, &mut app.tree);
    architecture.append(module2, &mut app.tree);

    app.root_id = Some(root);
    app.active_node_id = Some(root);

    app
}

#[test]
fn test_render_empty_mindmap() {
    let config = AppConfig::default();
    let mut app = AppState::new(config);

    let root = app.tree.new_node(Node::new("Empty Map".to_string()));
    app.root_id = Some(root);
    app.active_node_id = Some(root);

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_simple_tree() {
    let mut app = create_test_app_with_tree();

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_with_collapsed_node() {
    let mut app = create_test_app_with_tree();

    // Collapse the Features node
    let features_id = app.root_id.unwrap().children(&app.tree).next().unwrap();
    app.tree
        .get_mut(features_id)
        .unwrap()
        .get_mut()
        .is_collapsed = true;

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_with_active_node() {
    let mut app = create_test_app_with_tree();

    // Set active node to a child
    let features_id = app.root_id.unwrap().children(&app.tree).next().unwrap();
    app.active_node_id = Some(features_id);

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_edit_mode() {
    let mut app = create_test_app_with_tree();

    app.mode = AppMode::Editing {
        buffer: "Editing this node".to_string(),
        cursor_pos: 17,
    };

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_search_mode() {
    let mut app = create_test_app_with_tree();

    app.mode = AppMode::Search {
        query: "test search".to_string(),
    };

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_help_screen() {
    let mut app = create_test_app_with_tree();

    app.mode = AppMode::Help;

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_with_message() {
    let mut app = create_test_app_with_tree();

    app.set_message("File saved successfully!");

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_with_hidden_node() {
    let mut app = create_test_app_with_tree();

    // Mark a node as hidden
    let features_id = app.root_id.unwrap().children(&app.tree).next().unwrap();
    let task1_id = features_id.children(&app.tree).next().unwrap();
    app.tree.get_mut(task1_id).unwrap().get_mut().title = "[HIDDEN] Secret Task".to_string();

    // Hide hidden nodes
    app.config.show_hidden = false;

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_deep_tree() {
    let config = AppConfig::default();
    let mut app = AppState::new(config);

    // Create a deeper tree
    let root = app.tree.new_node(Node::new("Root".to_string()));
    let l1 = app.tree.new_node(Node::new("Level 1".to_string()));
    let l2 = app.tree.new_node(Node::new("Level 2".to_string()));
    let l3 = app.tree.new_node(Node::new("Level 3".to_string()));
    let l4 = app.tree.new_node(Node::new("Level 4".to_string()));

    root.append(l1, &mut app.tree);
    l1.append(l2, &mut app.tree);
    l2.append(l3, &mut app.tree);
    l3.append(l4, &mut app.tree);

    app.root_id = Some(root);
    app.active_node_id = Some(l3);

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_with_symbols() {
    let mut app = create_test_app_with_tree();

    // Add symbols to some nodes
    let root = app.root_id.unwrap();
    let features_id = root.children(&app.tree).next().unwrap();
    let task1 = features_id.children(&app.tree).next().unwrap();
    let task2 = features_id.children(&app.tree).nth(1).unwrap();

    // Mark first task as done (symbol1)
    app.tree.get_mut(task1).unwrap().get_mut().title =
        format!("{} Completed Task", app.config.symbol1);

    // Mark second task as failed (symbol2)
    app.tree.get_mut(task2).unwrap().get_mut().title =
        format!("{} Failed Task", app.config.symbol2);

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_wide_tree() {
    let config = AppConfig::default();
    let mut app = AppState::new(config);

    // Create tree with very long node titles
    let root = app.tree.new_node(Node::new(
        "This is a very long root node title that should wrap when displayed in the terminal".to_string()
    ));

    let child1 = app.tree.new_node(Node::new(
        "Another extremely long child node title that exceeds normal width constraints and needs proper text wrapping to display correctly".to_string()
    ));

    let child2 = app.tree.new_node(Node::new(
        "Short child".to_string()
    ));

    let grandchild = app.tree.new_node(Node::new(
        "This grandchild also has a considerably long title that will test the wrapping behavior at deeper levels of the tree structure".to_string()
    ));

    root.append(child1, &mut app.tree);
    root.append(child2, &mut app.tree);
    child1.append(grandchild, &mut app.tree);

    app.root_id = Some(root);
    app.active_node_id = Some(root);

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_very_deep_tree() {
    let config = AppConfig::default();
    let mut app = AppState::new(config);

    // Create a very deep tree (10+ levels)
    let mut current = app.tree.new_node(Node::new("Level 0 (Root)".to_string()));
    app.root_id = Some(current);
    app.active_node_id = Some(current);

    for i in 1..=10 {
        let child = app.tree.new_node(Node::new(format!("Level {}", i)));
        current.append(child, &mut app.tree);
        current = child;
    }

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}

#[test]
fn test_render_with_viewport_offset() {
    let mut app = create_test_app_with_tree();

    // Simulate viewport scrolled right
    app.viewport_left = 10.0;

    let backend = TestBackend::new(80, 20);
    let mut terminal = Terminal::new(backend).unwrap();

    terminal.draw(|frame| ui::render(frame, &mut app)).unwrap();

    assert_snapshot!(terminal.backend());
}
