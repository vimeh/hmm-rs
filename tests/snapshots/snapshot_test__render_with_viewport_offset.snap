---
source: tests/snapshot_test.rs
expression: terminal.backend()
---
"                                                                                "
"                       ╭✓ Completed Task                                        "
"         ╭Features ────┤                                                        "
"         │             ╰✗ Failed Task                                           "
"Root ────┤                                                                      "
"         │                 ╭model.rs                                            "
"         ╰Architecture ────┤                                                    "
"                           ╰ui.rs                                               "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"h-m-m | 7 nodes                                                                 "
